pkg/vm/engine/disttae/partition_state.go

```diff
@@ -0,0 +1,288 @@
+// limitations under the License.
+
+package disttae
+
+import (
+	"context"
+	"net/http"
+	"runtime/trace"
+
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/fileservice"
+	"github.com/matrixorigin/matrixone/pkg/pb/api"
+	"github.com/tidwall/btree"
+)
+
+var partitionStateProfileHandler = fileservice.NewProfileHandler()
+
+func init() {
+	http.Handle("/debug/cn-partition-state", partitionStateProfileHandler)
+}
+
+type PartitionState struct {
+	Rows   *btree.BTreeG[*RowEntry]
+	Blocks *btree.BTreeG[*BlockEntry]
+}
+
+type RowEntry struct {
+	BlockID uint64
+	RowID   types.Rowid
+	Time    types.TS
+
+	Deleted bool
+	Batch   *api.Batch
+	Offset  int
+}
+
+func (r *RowEntry) Less(than *RowEntry) bool {
+	if r.BlockID < than.BlockID {
+		return true
+	}
+	if than.BlockID < r.BlockID {
+		return false
+	}
+	if r.RowID.Less(than.RowID) {
+		return true
+	}
+	if than.RowID.Less(r.RowID) {
+		return false
+	}
+	if than.Time.Less(r.Time) {
+		return true
+	}
+	if r.Time.Less(than.Time) {
+		return false
+	}
+	return false
+}
+
+type BlockEntry struct {
+	BlockID uint64
+
+	MetaLocation  string
+	DeltaLocation string
+	SegmentID     uint64
+	Sorted        bool
+	CreateTime    types.TS
+	CommitTime    types.TS
+	DeleteTime    types.TS
+}
+
+func (b *BlockEntry) Less(than *BlockEntry) bool {
+	if b.BlockID < than.BlockID {
+		return true
+	}
+	if than.BlockID < b.BlockID {
+		return false
+	}
+	return false
+}
+
+func NewPartitionState() *PartitionState {
+	return &PartitionState{
+		Rows:   btree.NewBTreeG((*RowEntry).Less),
+		Blocks: btree.NewBTreeG((*BlockEntry).Less),
+	}
+}
+
+func (p *PartitionState) Copy() *PartitionState {
+	return &PartitionState{
+		Rows:   p.Rows.Copy(),
+		Blocks: p.Blocks.Copy(),
+	}
+}
+
+func (p *PartitionState) HandleLogtailEntry(ctx context.Context, entry *api.Entry) {
+	switch entry.EntryType {
+	case api.Entry_Insert:
+		if isMetaTable(entry.TableName) {
+			p.HandleMetadataInsert(ctx, entry.Bat)
+		} else {
+			p.HandleRowsInsert(ctx, entry.Bat)
+		}
+	case api.Entry_Delete:
+		if isMetaTable(entry.TableName) {
+			p.HandleMetadataDelete(ctx, entry.Bat)
+		} else {
+			p.HandleRowsDelete(ctx, entry.Bat)
+		}
+	default:
+		panic("unknown entry type")
+	}
+}
+
+func (p *PartitionState) HandleRowsInsert(ctx context.Context, input *api.Batch) {
+	ctx, task := trace.NewTask(ctx, "PartitionState.HandleRowsInsert")
+	defer task.End()
+
+	rowIDVector := vector.MustTCols[types.Rowid](mustVectorFromProto(input.Vecs[0]))
+	timeVector := vector.MustTCols[types.TS](mustVectorFromProto(input.Vecs[1]))
+
+	for i, rowID := range rowIDVector {
+		trace.WithRegion(ctx, "handle a row", func() {
+
+			blockID := blockIDFromRowID(rowID)
+			pivot := &RowEntry{
+				BlockID: blockID,
+				RowID:   rowID,
+				Time:    timeVector[i],
+			}
+			entry, ok := p.Rows.Get(pivot)
```

Comment:
Does the new implementation still need to be an upsert?

Reply:
Yes, in-place update.